pkg/tfbridge/validate_input_types.go

```diff
@@ -0,0 +1,361 @@
+	propertyValue resource.PropertyValue,
+	typeSpec pschema.TypeSpec,
+	propertyPath resource.PropertyPath,
+) []TypeFailure {
+	// don't type check
+	// - resource references (not yet)
+	// - assets (not yet)
+	// - archives (not yet)
+	// - computed values (they are allowed for any type)
+	// - unknown output values (similar to computed)
+	// - nulls (they are allowed for any type, and for missing required properties see validatePropertyMap)
+	if propertyValue.IsResourceReference() ||
+		propertyValue.IsAsset() ||
+		propertyValue.IsArchive() ||
+		propertyValue.IsComputed() ||
+		propertyValue.IsNull() ||
+		(propertyValue.IsOutput() && !propertyValue.OutputValue().Known) {
+		return nil
+	}
+
+	// for known first-class outputs simply validate their known value
+	if propertyValue.IsOutput() && propertyValue.OutputValue().Known {
+		elementValue := propertyValue.OutputValue().Element
+		return v.validatePropertyValue(elementValue, typeSpec, propertyPath)
+	}
+
+	// for secrets validate their inner value
+	if propertyValue.IsSecret() {
+		elementValue := propertyValue.SecretValue().Element
+		return v.validatePropertyValue(elementValue, typeSpec, propertyPath)
+	}
+
+	// now we are going to switch on the desired type
+	//
+	// a good reference here for the semantics of the TypeSpec is
+	// https://github.com/pulumi/pulumi/blob/master/pkg/codegen/schema/bind.go#L881
+	//
+	// this code is not using the binder functionality as that would require a plugin loader to resolve references,
+	// which is not desirable for the provider runtime
+	//
+	// possibly this could be done in the future with a way to still use the bind code while resolving references to
+	// unchecked types as using the bound representation would make the code cleaner
+	//
+	// for now the code simply follows bindSpec
+
+	if typeSpec.Ref != "" {
+		objType := v.getType(typeSpec.Ref)
+		// Refusing to validate unknown or unresolved type.
+		if objType == nil {
+			return nil
+		}
+
+		if !propertyValue.IsObject() {
+			return []TypeFailure{{
+				ResourcePath: propertyPath.String(),
+				Reason: fmt.Sprintf(
+					"expected object type, got %s type",
+					propertyValue.TypeString(),
+				),
+			}}
+		}
+
+		return v.validatePropertyMap(
+			propertyValue.ObjectValue(),
+			objType.Required,
+			objType.Properties,
+			propertyPath,
+		)
+	}
+
+	if typeSpec.OneOf != nil {
+		// bindTypeSpecOneOf provides a good hint of how to interpret these:
+		//
+		// https://github.com/pulumi/pulumi/blob/master/pkg/codegen/schema/bind.go#L842
+		//
+		// Specifically it defines the defaultType, discriminator, mapping and elements.
+		if len(typeSpec.OneOf) < 2 {
+			// invalid type, don't validate
+			return nil
+		}
+
+		// don't validate types with Discriminator (yet)
+		if typeSpec.Discriminator != nil {
+			return nil
+		}
+
+		// default type
+		if typeSpec.Type != "" {
+			typeSpec.OneOf = append(typeSpec.OneOf, pschema.TypeSpec{Type: typeSpec.Type})
```

Comment:
I always worry this mutation would leak outside the scope of this func, but looks like it doesn't perhaps because we have our own struct copy.

Reply:
I don't quite understand what this is doing. Why is `typeSpec.Type` appended to `OneOf`?

Reply:
I took that logic from https://github.com/pulumi/pulumi/blob/477a54b3de081a4927df3038c7b2266d599d5f3c/pkg/codegen/schema/bind.go#L849-L854

Not sure exactly why, but it looks like a "default" type can be stored in the `Type` field.

Reply:
This is a strange encoding I don't quite fully understand how it's interpreted either, would need to dig in.

Looking for examples in the wild.

https://github.com/pulumi/pulumi-aws/blob/upstream-v5.46.0/provider/cmd/pulumi-resource-aws/schema.json#L126765

Looks like for all examples I could find, these are enums that can also be strings. So the encoding is specifying that .type is string ("default"), but then also providing string as one of the alternatives, and the enum type as the other alternative.